pkg/services/auth/auth.go

```diff
@@ -0,0 +1,242 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/weaveworks/weave-gitops/pkg/flux"
+	"github.com/weaveworks/weave-gitops/pkg/git"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// BlockingCLIAuthHandler takes over the terminal experience and returns a token when the user completes the flow.
+type BlockingCLIAuthHandler func(context.Context, io.Writer) (string, error)
+
+func NewAuthCLIHandler(name gitproviders.GitProviderName) (BlockingCLIAuthHandler, error) {
+	switch name {
+	case gitproviders.GitProviderGitHub:
+		return NewGithubDeviceFlowHandler(http.DefaultClient), nil
+
+	}
+
+	return nil, fmt.Errorf("unsupported auth provider \"%s\"", name)
+}
+
+type AuthService struct {
+	logger     logger.Logger
+	fluxClient flux.Flux
+	// Note that this is a k8s go-client, NOT a wego kube.Kube interface.
+	// That interface wasn't providing any valuable abstraction for this service.
+	k8sClient   client.Client
+	gitProvider gitproviders.GitProvider
+	gitClient   git.Git
+}
+
+// NewAuthService constructs an auth service for doing git and GitProvider operations with authenticated clients.
+func NewAuthService(fluxClient flux.Flux, k8sClient client.Client, providerName gitproviders.GitProviderName, l logger.Logger, token string) (*AuthService, error) {
```

Comment:
I would make an interface for this and return that from this function instead of a pointer to the struct.  Albeit, the underlying object is still a pointer to a struct.  It's a pattern that need to start following across the codebase.

```
type AuthSomething interface {
	SetupDeployKey(ctx context.Context, name types.NamespacedName, targetName string, repo gitproviders.NormalizedRepoURL) error
	GitClient() (git.Git, error)
	GitProvider() gitproviders.GitProvider
}
```

Reply:
I had thought about an `interface` but didn't anticipate this needing to be mocked, since it provides an existing "client" `interface` in `.GitClient()`.

I think I will refactor back to an `interface`, since some future tests might want to mock it.